keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,438 @@
+    type PodIPInfo struct {
+        IP string
+        Properties map[string]string
+    }
+
+    // IP addresses allocated to the pod with associated metadata.
+    PodIPs []PodIPInfo `json:"podIPs,omitempty" protobuf:"bytes,6,opt,name=podIPs"`
+```
+This internal representation should eventually become part of a versioned API (after a period of deprecation for the singular "PodIP" field).
+
+#### kubelet Startup Configuration for Dual-Stack Pod CIDRs
+The existing "--pod-cidr" option for the [kubelet startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) will be modified to support multiple IP CIDRs in a comma-separated list (rather than a single IP string), i.e.:
+```
+  --pod-cidr  ipNetSlice   [IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+Only the first address of each IP family will be used; all others will be ignored.
+
+#### kube-proxy Startup Configuration for Dual-Stack Pod CIDRs
+The existing "cluster-cidr" option for the [kube-proxy startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) will be modified to support multiple cluster CIDRs in a comma-separated list (rather than a single IP string), i.e:
+```
+  --cluster-cidr  ipNetSlice   [IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+Only the first address of each IP family will be used; all others will be ignored.
+
+#### 'kubectl get pods -o wide' Command Display for Dual-Stack Pod Addresses
+The output for the 'kubectl get pods -o wide' command will need to be modified to display a comma-separated list of IPs for each pod, e.g.:
+```
+       kube-master# kubectl get pods -o wide
+       NAME               READY     STATUS    RESTARTS   AGE       IP                          NODE
+       nginx-controller   1/1       Running   0          20m       fd00:db8:1::2,192.168.1.3   kube-minion-1
+       kube-master#
+```
+
+#### 'kubectl describe pod ...' Command Display for Dual-Stack Pod Addresses
+The output for the 'kubectl describe pod ...' command will need to be modified to display a comma-separated list of IPs for each pod, e.g.:
+```
+       kube-master# kubectl describe pod nginx-controller
+       .
+       .
+       .
+       IPs:     fd00:db8:1::2,192.168.1.3
+       .
+       .
+       .
+```
+
+### Container Networking Interface (CNI) Plugin Considerations
+
+This feature requires the use of CNI API 0.3.1 or later. The dual-stack feature requires no changes to this API.
+
+### Configuration of IP Family in Service Definitions
+This proposal adds an option to configure an IP family for each port configuration within each service definition.
+```
+    ipFamily: <ipv4|ipv6|dual-stack>       [Default: dual-stack]
+```
+For example, a ports definition for an application that only binds to IPv4 might look like this:
+```
+  ports:
+  - port: 8080
+    nodePort: 30302
+    targetPort: 80
+    protocol: TCP
+    ipFamily: ipv4
```

Comment:
I expect to eventually have to match other address/network properties here.

Reply:
Can you clarify what you mean here?